CBMC: Add proof and contract for polyveck_uniform_eta

jakemas opened the pull request:
- Resolves https://github.com/pq-code-package/mldsa-native/issues/123
- Depends on #224

mkannwischer commented on mldsa/polyvec.c:
```diff
@@ -201,16 +201,37 @@ int polyvecl_chknorm(const polyvecl *v, int32_t bound)
+    poly t;
+    poly_uniform_eta(&t, seed, n);
+
+#ifdef CBMC
+#pragma CPROVER check push
+#pragma CPROVER check disable "unsigned-overflow"
+#endif
+    n++;
+#ifdef CBMC
+#pragma CPROVER check pop
+#endif
```
The function is only called with nonce=L`, so this can never overflow in practise. I'd prefer a stronger pre-condition over silencing the warning here.

mkannwischer reviewed (changes requested):
Thanks @jakemas. One comment, rest looks good.

hanno-becker commented on mldsa/polyvec.c:
```diff
@@ -39,13 +39,33 @@ void polyvec_matrix_pointwise_montgomery(polyveck *t,
+    poly t;
+    poly_uniform_eta(&t, seed, n);
+
+#ifdef CBMC
+#pragma CPROVER check push
+#pragma CPROVER check disable "unsigned-overflow"
+#endif
+    n++;
+#ifdef CBMC
+#pragma CPROVER check pop
+#endif
```
As in #229, this should be replaced by a precondition bound on `nonce`.

jakemas replied:
Ignore this, this PR is just staged ontop of the old one that has now been updated. This will be removed when I rebase

hanno-becker replied:
Ah right, this was from the old PR -- I misread!

hanno-becker reviewed (changes requested):
We should add a precondition bound for `nonce` rather than disabling the overflow check.

@jakemas Please don't disable any standard macros without explicit discussion and comment in the code -- this is dangerous and undermines the claims we are making about the scope of the CBMC proofs.

jakemas commented:
> We should add a precondition bound for `nonce` rather than disabling the overflow check. We have to be careful about changing the default set of CBMC checks, as it can render our messaging about the scope of CBMC inaccurate.

I did! I think you were reviewing `polyvecl_uniform_eta`. I rebased ontop of main now `poly_uniform_eta` is merged, and removed all other commits to remove this confusion.

hanno-becker commented on mldsa/polyvec.h:
```diff
@@ -165,8 +165,29 @@ typedef struct
+__contract__(
+  requires(memory_no_alias(v, sizeof(polyveck)))
+  requires(memory_no_alias(seed, MLDSA_CRHBYTES))
+  requires(nonce == MLDSA_L)
```
Can we have  `<= UINT16_MAX - MLDSA_L` here, as in #229?

jakemas replied:
fixed

hanno-becker reviewed (approved):
Thanks @jakemas !

mkannwischer reviewed (approved):
LGTM. Thanks for the changes @jakemas!
